### ISSUE TITLE
IPC leader does not wait for all processes to finish

### ISSUE THREAD

**Issue by boryspoplawski** [oscarlab/graphene#2514]
## Description of the problem
If the first process, which is the IPC leader, exits, all other processes cannot use IPC any more. We need to decide on a first process exit policy and implement it. I see two options:
- IPC leader kills all other processes before exiting
- IPC leader just cleans up app resources, but stays alive as long as other processes are still running.

We already have a TODO about it:
https://github.com/oscarlab/graphene/blob/4f8d6fb8b323d62f00be2ac2c7239878b14afc1e/LibOS/shim/src/sys/shim_exit.c#L23-L27

**Comment by mkow** [comment]
The second option seems better to me (I think it's fine if the main process of a user app just spawns something and then exits right away leaving the child).

**Comment by dimakuv** [comment]
> (I think it's fine if the main process of a user app just spawns something and then exits right away leaving the child)

But then Graphene will have unexpected behavior: the caller of this user app expects the main process to exit immediately, but instead the caller will hang because the main process is still alive. E.g., when we start Memcached in daemon mode, we expect `memcached -d` to return immediately, and we continue with our Bash session, but with this second option, the process will simply live forever (until `kill` on its child). This sounds like a surprising side effect for end users.

**Comment by boryspoplawski** [comment]
@dimakuv Well, that call would have `graphene-something` prepended so it wouldn't be *that* surprising. Also going with the first option would just... kill the memcached.

**Comment by dimakuv** [comment]
Yes, both options don't work for this daemon case. I was thinking about "electing a new IPC leader" third option, but this sounds... complex.

If choosing between these two options, I also vote for the second option.

**Comment by boryspoplawski** [comment]
> Yes, both options don't work for this daemon case.

It does not work in the sense it will stay up and not become a "real daemon". But I think if you want to have Graphene as a daemon process on the host, then you should wrap it in some script doing that. Or, in some future, we could add a `-d` flag to Graphene invocation which would daemonize Graphene as a whole.

**Comment by dimakuv** [comment]
A list of related issues that have this "IPC leader does not wait for all processes to finish" bug:
- https://github.com/oscarlab/graphene/issues/2426
- https://github.com/oscarlab/graphene/issues/2531
- https://github.com/oscarlab/graphene/issues/2541
- https://github.com/oscarlab/graphene/issues/2542

We closed these issues since they are tracked here. Any new issues that boil down to this bug will be closed and mentioned in this meta-issue. The closing of other issues is to reduce the number of open issues, for easier management. (In other words, it doesn't invalidate the closed issues, it's just that it's very hard to navigate through a huge list of issues.)

**Comment by boryspoplawski** [comment]
@dimakuv wait, most of the linked issues exhibit the bug described here, but this bug is not the main reason they fail (it just causes some error messages, but does not make the test fail - that happen before these error msgs). #2541 is a good example

**Comment by anjalirai-intel** [comment]
Hi @dimakuv 

I am attaching the list of tests which fails with IPC issue here for the verification purpose. Some of these tests are not placed in original bug.

- [ ] waitpid04 
- [ ] waitpid03 
- [ ] fcntl12 
- [ ] recvfrom01 
- [ ] send01 
- [ ] connect01 
- [ ] recv01 
- [ ] sendto01 
- [ ] fcntl12_64 
- [ ] kill09

**Comment by boryspoplawski** [comment]
What is worth noting is that these tests do not fail, they just print an error message at the end, but still succeed.

**Comment by boryspoplawski** [comment]
> I am attaching the list of tests which fails with IPC issue here for the verification purpose. Some of these tests are not placed in original bug.

I've verified that list - non of these fail due to the problem described in this issue. Some of them do sometimes print an error message described here, but it happens already after the test finished and is unrelated to actual result of the test (be it fail or success).

**Comment by boryspoplawski** [comment]
On the original topic: I don't think implementing 2nd option will be that easy: Graphene does not bother cleaning most of the stuff up at exit (depending on host to clean it, which is not a problem atm), so this would require a lot of changes (e.g. closing all files, releasing all handles, cleaning up all threads and all user memory, lot of other cleanups).
